blog/content/post/2022-04-12-talos2_2nd_cpu_and_testing.md

```diff
@@ -0,0 +1,93 @@
+small PowerPC-lite core inside POWER9 SoC), now we have to do something similar
+for the other processor. Before that happens, all SCOM accesses are done through
+FSI (Flexible Service Interface).
+
+This added another layer in accessing buses that are already accessed through
+SCOM. An example of such bus is I2C, which is used (among other uses) for
+accessing MVPD (Module VPD). MVPD holds important data about CPU, like its
+operating frequencies and voltages or list of enabled cores. It must be read
+before XSCOM is enabled, meaning it has to happen using slow FSI communication.
+How slow are we talking about? Reading 64 KB of MVPD takes 4 seconds when done
+through XSCOM (primary CPU), but 9 seconds over FSI (secondary CPU). It is used
+both in romstage and ramstage, so we quickly decided it is worth to cache MVPD
+between stages. Next step will be saving it to flash and read it through I2C
+only if necessary, i.e. after flashing whole PNOR or when processor was swapped
+and saved image doesn't match hardware. Hostboot does it this way, so should we.
+This could save about 13 seconds of boot time.
+
+Next item on the list was enabling and training X bus links. An X bus is the
+socket-to-socket SMP interconnect between two POWER9 processors. Before this
+step each CPU worked independently of the other one, in so-called island mode.
+In that mode it would be impossible to directly access address spaces (memory
+and MMIO) behind remote CPU. Enabling SMP makes XSCOM access to secondary
+processor's devices possible.
+
+With SCOM, I2C and X Bus out of the way, most of initialization consisted of
+repeating existing steps for the second CPU. There were some exceptions where
+simple repetition was not enough, most notably:
+
+- TOD (time of day) - in case of discrepancy between two clocks one of them
+  takes precedence in setting the other. Complex calculations are done during
+  the initialization to choose "more important" clock source - the one with
+  higher delay, as it is more inert.
+- memory controller - even though memory initialization and training is repeated
+  for secondary processor, only main CPU sets a common memory space address
+  translation maps.
+- PCIe controller - links are split differently between ports connected to each
+  of processor sockets.
+
+In addition to various initialization steps, we had to update device tree passed
+to Skiboot to include information about new CPU and devices it adds. We also
+switched away from FIT payload and started using ELF. Main motivation for this
+change was fact that [FIT decompression code does not work as expected](https://mail.coreboot.org/hyperkitty/list/coreboot@coreboot.org/thread/6EZWU7YPUJE564GNCV7U32IXWPFTV7FB/)
+in coreboot. With payload compressed with LZMA, flash memory footprint of whole
+coreboot and Skiboot has shrunk to just below 700 KB. For comparison, Hostboot
+takes about 30 MB of space.
+
+## New tests
+
+TBD
```

Comment:
@sulewskiprzemyslaw We would expect short description of changes in the test area here.